src/librustc_trans/back/link.rs

```diff
@@ -146,9 +146,7 @@ pub(crate) fn link_binary(sess: &Session,
     let mut out_filenames = Vec::new();
     for &crate_type in sess.crate_types.borrow().iter() {
         // Ignore executable crates if we have -Z no-trans, as they will error.
-        if (sess.opts.debugging_opts.no_trans ||
-            !sess.opts.output_types.should_trans()) &&
-           crate_type == config::CrateTypeExecutable {
+        if sess.opts.debugging_opts.no_trans && crate_type == config::CrateTypeExecutable {
```

Comment:
I'm not sure that this is correct. This loop goes on to link the code units which have been generated, but if we are in `!sess.opts.output_types.should_trans()` mode, then there should be no code units to link. Also I think it is odd that `emit=metadata` and `-Ztrans` would have different behaviour, which is what is happening now.

However, it has been a long time since I wrote this, so I'm not really sure what *should* be done, sorry

Reply:
Would you find it acceptable just to special-case `!sess.opts.output_types.should_trans() || sess.opts.output_types.contains(&OutputType::Metadata)` as a quick fix for now?

Reply:
I'm really not sure, sorry. ping @rust-lang/compiler anybody know about this code?